src/utils.js

```diff
@@ -31,3 +31,60 @@ export function createColorGenerator(min_value, max_value, color) {
   }
   return colorGenerator;
 }
+
+// Create 'summary' dictionary: Record<string, {name: string; value: number}>
+function createSummaryDictionary(summariesOrDetails) {
+  return summariesOrDetails.reduce((acc, curr) => {
+    if (acc[curr.name] === undefined) {
+      acc[curr.name] = {
+        name: curr.name,
+        value: curr.value,
+      };
+    } else {
+      acc[curr.name].value += curr.value;
+    }
+    return acc;
+  }, {});
+}
+
+// Sort summary dictionary
+function sortSummaryDictionary(summaryDictionary) {
+  return Object.values(summaryDictionary).sort((a, b) => {
+    return b.value - a.value;
+  });
+}
+
+export function getYearSummary(data, date) {
+  let summaries = [];
+  /**
+   * Filtering the 'data' based on the year of date,
+   * extracting all the summaries from the data into a single dimensional array and
+   */
+  if (Array.isArray(data)) {
+    const summaries = data
+      .filter(
+        (e) => new Date(e.date).getFullYear() === new Date(date).getFullYear()
+      )
+      .flatMap((e) => e.summary);
+    const summaryDictionary = createSummaryDictionary(summaries);
+    summaries = sortSummaryDictionary(summaryDictionary);
+  }
+  return summaries;
+}
+
+// Calculate daily summary if that was not provided
+export function addSummary(data) {
+  let summaries = [];
+  if (Array.isArray(data)) {
+    if (data[0].summary === null || data[0].summary === undefined) {
+      summaries = data.map((d) => {
+        // Create project dictionary: Record<string, {name: string; value: number}>
+        const summaryDictionary = createSummaryDictionary(d.details);
+        // Update "summary" property of the array element
+        const summary = sortSummaryDictionary(summaryDictionary);
+        return summary;
+      });
+    }
+  }
+  return summaries;
+}
```

Comment:
is this acceptable sir ?

Reply:
i don't think so

how will attach the summary object to the corresponding datum object? this approach might make it difficult